src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -68,26 +64,57 @@ private StringComparison Comparison
             get { return CaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase; }
         }
 
-        protected override bool TokenKindEquals(SyntaxKind x, SyntaxKind y) => x == y;
-
-        public override void BuildSpan(SpanBuilder span, SourceLocation start, string content)
+        // Special tags include <!--, <!DOCTYPE, <![CDATA and <? tags
+        private bool AtSpecialTag
         {
-            span.Kind = SpanKindInternal.Markup;
-            span.ChunkGenerator = new MarkupChunkGenerator();
-            base.BuildSpan(span, start, content);
+            get
+            {
+                if (At(SyntaxKind.OpenAngle))
+                {
+                    if (NextIs(SyntaxKind.Bang))
+                    {
+                        return !IsBangEscape(lookahead: 1);
+                    }
+
+                    return NextIs(SyntaxKind.QuestionMark);
+                }
+
+                return false;
+            }
         }
 
-        protected override void OutputSpanBeforeRazorComment()
+        public RazorDocumentSyntax ParseDocument()
         {
-            Output(SpanKindInternal.Markup);
+            if (Context == null)
+            {
+                throw new InvalidOperationException(Resources.Parser_Context_Not_Set);
+            }
+
+            using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+            using (PushSpanContextConfig(DefaultMarkupSpanContext))
+            {
+                var builder = pooledResult.Builder;
+                NextToken();
+                while (!EndOfFile)
+                {
+                    SkipToAndParseCode(builder, SyntaxKind.OpenAngle);
+                    ParseTagInDocumentContext(builder);
+                }
+                AcceptMarkerTokenIfNecessary();
+                builder.Add(OutputAsMarkupLiteral());
+
+                var markup = SyntaxFactory.MarkupBlock(builder.ToList());
+
+                return SyntaxFactory.RazorDocument(markup);
+            }
         }
 
-        protected void SkipToAndParseCode(SyntaxKind type)
+        private void SkipToAndParseCode(in SyntaxListBuilder<RazorSyntaxNode> builder, SyntaxKind type)
```

Comment:
I assume once we evolve this class that SkipToAndParseCode goes away yes? Having it introduces too much genericness into our HtmlParser for us to properly evolve it in the future.

Reply:
Yes absolutely. https://github.com/aspnet/Razor/issues/2583 requires me to change most of these. I kept this to not introduce too much churn when we switched baselines for the first time.